src/schema.ts

```diff
@@ -0,0 +1,385 @@
+	tick: number;
+}
+/** Represents the dropping of a single bass marble. */
+export interface BassDropEvent {
+	kind: "bass";
+	/** The string (not note) to drop the marble onto. */
+	string: BassString;
+	/**
+	 * The fret to be played when this marble hits.
+	 *
+	 * Even though this technically requires user input, we opted to specify the fret here
+	 * because it's much simpler than requiring a separate fret pressed event in the program.
+	 */
+	fret: number;
+}
+/** Represents the dropping of a single drum marble. */
+export interface DrumDropEvent {
+	kind: "drum";
+	/** The drum to drop the marble onto. */
+	drum: DrumType;
+}
+/** Represents the dropping of a single vibraphone marble. */
+export interface VibraphoneDropEvent {
+	kind: "vibraphone";
+	/** The channel (key) to drop the marble onto.
+	 *
+	 * To figure out which note it will play, the [[VibraphoneState.notes]]
+	 * property must be referenced.
+	 * ```typescript
+	 * const vDropEvent: VibraphoneDropEvent;
+	 * const vState: VibraphoneState;
+	 * const noteHit = vState.notes[vDropEvent.channel];
+	 * ```
+	 */
+	channel: VibraphoneChannel;
+}
+
+/**
+ * Represents a single performance of the associated [[Program]].
+ */
+export interface Performance {
+	/** Information about this performance. */
+	metadata: PerformanceMetadata;
+	/** The program associated with this performance. */
+	program: Program;
+	/**
+	 * The state of the machine at the start of the performance.
+	 *
+	 * Note that this should not be changed during the performance.
+	 * Use [[Program.state]] if you want to keep track of the current
+	 * state of the machine. This means that [[Performance.initialState]]
+	 * should be copied into [[Program.state]] at tick 0 of Performance playback.
+	 */
+	readonly initialState: State;
+	/**
+	 * The events that make up this performance. As with [[Program.dropEvents]],
+	 * these events must be in ascending order.
+	 */
+	events: TimedEvent[];
+}
+/** Metadata for performance */
+export interface PerformanceMetadata {
+	/**
+	 * The title of the performance.
+	 * This doesn't have to be the same as the title of the program.
+	 */
+	title: string;
+	/**
+	 * The author/performer of the performance.
+	 * This doesn't have to be the same as the author of the program.
+	 */
+	author: string;
+}
+
+/**
+ * An untimed event that changes the state of one channel
+ * or manually drops a marble.
+ *
+ * Events only change one thing. Multiple simultaneous changes must
+ * be represented by multiple events with the same tick.
+ */
+export type Event =
+	| DropEvent
+	| MachineEvent
+	| VibraphoneEvent
+	| HihatMachineEvent
+	| HihatEvent
+	| BassEvent;
+
+/** Represents an event occurring in time (non-tempo dependent). */
+export type TimedEvent = BaseTimedEvent & Event;
+/** Represents information associated with all timed events. */
+export interface BaseTimedEvent {
+	/** The time (in seconds) that this event occurs. */
+	time: number;
+}
```

Comment:
I got really confused here for a while thinking that it was referring to the bass, so maybe a quick note that says "Not Bass instrument - Base absolute time" or something to that effect?

Reply:
Hmm, maybe we should just change the name. There's `CoreDropEvent` and this. There's no semantic reason to name them differently, right?

Reply:
This is addressed in #14.